ENH: Add random_state to multivariate distributions

FlorianWilhelm opened the pull request:
Add a custom `random_state` to multivariate distributions as it has been done for univariate distributions in PR #4129.

FlorianWilhelm commented on scipy/stats/_multivariate.py:
```diff
@@ -280,7 +281,7 @@ class multivariate_normal_gen(object):
     logpdf(x, mean=None, cov=1, allow_singular=False)
         Log of the probability density function.
-    rvs(mean=None, cov=1, allow_singular=False, size=1)
+    rvs(mean=None, cov=1, size=1, random_state=None)
```
@ev-br I just removed the `allow_singular` keyword because it does not seem to be used anymore.

FlorianWilhelm commented on scipy/stats/_multivariate.py:
```diff
@@ -847,16 +904,20 @@ def rvs(self, alpha, size=1):
         alpha = _dirichlet_check_parameters(alpha)
-        return np.random.dirichlet(alpha, size=size)
+        if random_state is None:
+            return self.random_state.dirichlet(alpha, size=size)
```
@ev-br Not sure I understand, self.random_state calls the property `random_state` which returns the private `self._random_state`. Do you want me to directly access `self._random_state`?

ev-br replied:
Yes.
Correct me if I'm wrong: using the property here incurs an overhead of `__get__`, which can be avoided. 
Granted, this is a minor quibble.

FlorianWilhelm replied:
@ev-br okay, I'll remove it

FlorianWilhelm commented:
@ev-br One more thing before I rebase, do you think it would be better to have an abstact `multi_rv_generic` just like `rv_generic` in `_distn_infrastructure`? It could hold the the random_state properties which are code duplications right now in four distributions.

FlorianWilhelm commented:
Rebased and forced update

ev-br commented:
Re multi_rv_generic: nice idea! There is probably no point adding anything else to the base class, not yet at least. Care to tackle it?

ev-br commented on scipy/stats/_multivariate.py:
```diff
@@ -1332,6 +1413,10 @@ def _standard_rvs(self, n, shape, dim, df):
+        random_state : None or int or np.random.RandomState instance, optional
+            If int or RandomState, use it for drawing the random variates.
+            If None (or np.random), the global np.random state is used.
+            Default is None.
```
This docstring is not quite right here: the random_state parameter should be an actual RandomState, no? The int-or-None game is played at the level of the public RVs method.

FlorianWilhelm commented:
@ev-br Will do tomorrow and ping you when done.

FlorianWilhelm commented:
@ev-br I introduced now `multi_rv_generic` and `multi_rv_frozen`. Besides that the `rvs` method of the frozen distributions now supports custom random_states too. Although, while looking through `_distn_infrastructure.py` I found a small bug that I corrected in commit fb4477608fc00f58b172311d5d9013ec8babd0ee.

ev-br commented on scipy/stats/_multivariate.py:
```diff
@@ -520,19 +582,19 @@ def __init__(self, mean=None, cov=1, allow_singular=False):
         self.cov_info = _PSD(self.cov, allow_singular=allow_singular)
-        self._mnorm = multivariate_normal_gen()
+        self.dist = multivariate_normal_gen(seed)
 
```
This technically makes `self.dist` a public attribute since it does not have a leading underscore. Could you revert this change here and elsewhere? I'd think we do not need to emulate a legacy decision made for univariate frozen distributions.

ev-br commented:
Looks pretty good to me. There are two more DRY fixes which may be either a part of this PR or be left for later. First, the documentation of the random_state parameter is repeated several times; it can be moved into a default parameters dictionary. Second, each `rvs` method repeats the same if-then-else block `if random_state is None...`. This can be moved to either a private helper of the superclass, or just be made a standalone helper function.

Like I'm saying, these can be left to a future PR. The change adding a public `dist` attribute should be reverted though.

Finally, you may want to squash the commits a bit (this is optional as well).

FlorianWilhelm commented:
@ev-br I moved the DRY if-then-else random_state into `multi_rv_generic`. Also I tried to move the random_state doc into this doccer dictionary. Since I have never done this before, can you please check if it is correct? 
If it is okay for you I would not want to squash my commits. Somehow I don't like the idea of tampering with commits that represent a sequence of design decisions and changes throughout time that eventually led to a solution.

ev-br commented:
OK, docstring generation is a bit of a mess. I've added a small tweak, can you cherry-pick the last commit from https://github.com/ev-br/scipy/tree/pr/4318
With that last one, the random_state docs show in the html documentation (to build the docs locally, I use `python setup.py install; cd doc; make html-scipyorg` --- this might or might not be the best way)
Inspecting the docstrings if live objects shows that there are some bare keys left --- try 
```
>>> from scipy.stats import multivariate_normal as mvn
>>> help(mvn)
```
but this is a separate, preexisting problem which can be taken care of separately.

With that last tweak I think this PR is done. I'll keep it open for a while in case somebody else wants to comment.

FlorianWilhelm commented:
@ev-br okay, cherry-picked your commit. Thank you.
I can confirm the problems with the docstrings of live objects but I have no clue how to solve it, so I would agree that it is a good idea to leave it to someone else :-)

ev-br commented:
Ouch. 
Idiotically, I left out one other obscure problem with the docstrings. In the `Methods` section, having a `random_state` parameter confuses Sphinx because the underscore has special meaning in Sphinx. 

I cannot find the exact issue/PR right now, but from memory, a similar problem was fixed for univariate distributions this August by surrounding the signatures of the method by double backticks. 
If you feel like fixing this, great. Otherwise, I'll do it later.

ev-br commented:
Re: squashing the commits vs preserving the history. In a project of scipy size, it's more about features rather than particular details of how the implementation evolved in time. For example, What is the value preserving the fact that we did not immediately realize that underscores need to be shielded by double backticks in docstrings?
There is no hard and fast rule for this of course, but leaving 16 commits does look a bit excessive to me.

Also, you should add yourself to THANKS.txt if you did not do it yet.

FlorianWilhelm commented:
@ev-br Okay, I agree, just squashed most of the commits and added myself to THANKS.txt.

ev-br commented:
Merged, thanks @FlorianWilhelm!
